lib/explorer/backend/data_frame.ex

```diff
@@ -59,6 +59,11 @@ defmodule Explorer.Backend.DataFrame do
   @callback shape(df) :: {integer(), integer()}
   @callback n_rows(df) :: integer()
   @callback n_columns(df) :: integer()
+  @callback data_for_inspect(df, opts :: Inspect.Opts.t()) ::
+              {rows :: nil | non_neg_integer(),
+               columns ::
+                 non_neg_integer(), backend :: atom() | String.t(),
+               series :: [{name :: String.t(), dtype :: atom() | String.t(), values :: list()}]}
```

Comment:
I think this contract is a bit too constrained. For example, in ExplorerSQL, we may want to show the SQL query when inspecting. Perhaps we could return this:

    {backend :: string,
     rows :: nil | non_neg_integer,
     cols :: nil | non_neg_integer,
     prelude :: Inspect.Algebra.t,
     series :: ...}

Some notes:

  * `prelude` is an algebra document that is placed right below `Backend[rows x cols]`. Most backends can default to `empty()`.

  * I have said that `cols` can be nil based on a conversation with @philss. For example, should `ExplorerSQL` do a query to get the table fields+types as soon as you say the table? Or should we say we simply don't know the table information until when the query is made? We don't have to support this but we might as well make it consistent with rows?

  * I think you can also force more strict types here. Such as requiring the `backend` to be a string and the dtype to be an atom (which is its normal representation)

Reply:
I think with SQL it will be quite unwieldy to display the SQL query when inspecting. Why not just allow the user to call for it with a function? But that's predicated on: if it's SQL I'd want to fetch the query with a limit (`opts.limit + 1`), like Polars lazy. This is what `dbplyr` does.

Reply:
In `dbplyr` they have `show_query`:

```
> show_query(mtcars2)
<SQL>
SELECT *
FROM `mtcars`
```

Reply:
Yeah, I think we will not end up showing the underlying SQL, but I thought having the provision for such a mechanism is good in case we need it in the future? Otherwise we will have to enlarge/deprecate the behaviour function if/when we need it.